Add frame Copier

lherman-cs opened the pull request:
This is an incremental effort to redesign the project. Frame copy is necessary for supporting a video broadcaster.

Benchmark:

```
goos: darwin
goarch: amd64
pkg: github.com/pion/mediadevices/pkg/io/video
BenchmarkCopyOptimized
BenchmarkCopyOptimized-4            9290            132430 ns/op             670 B/op          0 allocs/op
BenchmarkCopyNaive
BenchmarkCopyNaive-4                2613            446601 ns/op         3130664 B/op          4 allocs/op
PASS
ok      github.com/pion/mediadevices/pkg/io/video       5.769s
```

at-wat commented on pkg/io/video/copy.go:
```diff
@@ -0,0 +1,196 @@
+
+// Copy copies src to dst. Since dst can be nil, the caller is responsible to assign the
+// return value from Copy to dst. When dst is not nil, Copy will reuse its memory and only
+// copy values from src to dst.
```
I think it's better to split shallow copy func and deep copy func with explicit names.

lherman-cs replied:
Not sure what you mean, but this function only does deep copy.

at-wat replied:
> When dst is not nil, Copy will reuse its memory and only copy values from src to dst.

This function looks doing deep copy only if dst is not nil.

lherman-cs replied:
Oh I see what you mean. So, I guess you're thinking about users misusing this by calling copy from the same Copier instance in different goroutines. I can definitely see that to be problematic.

lherman-cs replied:
So, I think we'll create `ShallowCopy` and `DeepCopy`. But, should we allow `ShallowCopy` to have `nil` as its dst? Or should we just enforce users to not have `nil` at all as its dst?

at-wat commented on pkg/io/video/copy.go:
```diff
@@ -0,0 +1,196 @@
+			*clone = *src
+		} else {
+			copied := *src
+			clone = &copied
```
I feel it's better to return error if type of src and dst is not matched.

lherman-cs replied:
Why is that? The thing is if `dst` is nil, ok will be equal to false anyway. So, I think it'll only add extra codes and more overhead.

at-wat replied:
I think if user try to copy `*image.CMYK` into `*image.RGBA`, it's misuse of the function.

lherman-cs replied:
Yep, makes sense. I think we'll just enforce this with panic since this is a misuse.

lherman-cs replied:
After second thought, I'd like to keep the behaviour as it is. My motivation is that I want to use Copy for another function, "broadcast". So, it's important to keep it generic since users might change the image format at anytime in the pipeline.

lherman-cs commented:
@at-wat I think the naming is really confusing. I've updated by renaming it to `FrameBuffer` instead, and changed the API to be `Load` and `Store` style. This way, it'll let the users aware the buffer is not meant to be shared across multiple goroutines unless there's a synchronization in place.

The design is inspired by atomic.Value, https://golang.org/pkg/sync/atomic/#Value.Load.

What do you think about this approach?

at-wat commented:
`FrameBuffer` API sounds good to me!

lherman-cs commented:
@at-wat updated the PR, PTAL.

at-wat reviewed (approved):
LGTM
